### PR TITLE
Dispatch take_along_axis to gather

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107689
* #107746
* __->__ #107711
* #107710
* #107688
* #107687

Gather does the same thing, but it's much better supported in the
`torch.compile` stack

cc @mruberry @rgommers @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
tests?

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -2413,10 +2413,6 @@ def sample_inputs_gather(op_info, device, dtype, requires_grad, **kwargs):
         make_arg((M, S)),
         1,
         gather_variable((M, S // 2), 0, S, True, device=device))
-    yield SampleInput(
-        make_arg(),
-        0,
-        torch.tensor([0], dtype=torch.int64, device=device))
```
Comment (lezcano) [inline torch/testing/_internal/common_methods_invocations.py:2419]: Repeated sample.

Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:2419]: @lezcano one of these is `torch.tensor([0])` (i.e. 1d) and the other is `torch.tensor(0)` (i.e. 0d)

Comment (lezcano) [reply torch/testing/_internal/common_methods_invocations.py:2419]: Ah, right, sorry for that. Reverting at https://github.com/pytorch/pytorch/pull/107776

**Comment by ev-br** [comment]
This seems to fail on main locally:

```
$ pytest test/torch_np/numpy_tests/lib/test_shape_base_.py
Test session starts (platform: linux, Python 3.8.17, pytest 7.4.0, pytest-sugar 0.9.7)
rootdir: /home/ev-br/repos/pytorch
configfile: pytest.ini
plugins: hypothesis-6.82.6, sugar-0.9.7, xdist-3.3.1
collected 72 items                                                                                                                                                                                        

 test/torch_np/numpy_tests/lib/test_shape_base_.py ✓✓✓                                                                                                                                        4% ▌         

―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― TestTakeAlongAxis.test_broadcast ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Traceback (most recent call last):
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/_pytest/runner.py", line 341, in from_call
    result: Optional[TResult] = func()
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/_pytest/runner.py", line 262, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_callers.py", line 155, in _multicall
    return outcome.get_result()
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_result.py", line 108, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/_pytest/runner.py", line 177, in pytest_runtest_call
    raise e
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/_pytest/runner.py", line 169, in pytest_runtest_call
    item.runtest()
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/_pytest/python.py", line 1788, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_callers.py", line 116, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/ev-br/.conda/envs/pytorch-dev/lib/python3.8/site-packages/_pytest/python.py", line 194, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/home/ev-br/repos/pytorch/test/torch_np/numpy_tests/lib/test_shape_base_.py", line 101, in test_broadcast
    actual = take_along_axis(a, ai, axis=1)
  File "/home/ev-br/repos/pytorch/torch/_numpy/_normalizations.py", line 213, in wrapped
    result = func(*args, **kwds)
  File "/home/ev-br/repos/pytorch/torch/_numpy/_funcs_impl.py", line 889, in take_along_axis
    return torch.gather(arr, axis, indices)
RuntimeError: Size does not match at dimension 2 expected index [1, 2, 5] to be smaller than self [3, 4, 1] apart from dimension 1

 test/torch_np/numpy_tests/lib/test_shape_base_.py ⨯✓x✓✓xxxxxxxxx✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓sx✓✓✓✓✓✓✓✓✓✓✓x                                                                    100% ██████████
========================================================================================= short test summary info =========================================================================================
FAILED [0.0032s] test/torch_np/numpy_tests/lib/test_shape_base_.py::TestTakeAlongAxis::test_broadcast - RuntimeError: Size does not match at dimension 2 expected index [1, 2, 5] to be smaller than self [3, 4, 1] apart from dimension 1

Results (0.51s):
      58 passed
       1 failed
         - test/torch_np/numpy_tests/lib/test_shape_base_.py:97 TestTakeAlongAxis.test_broadcast
      12 xfailed
       1 skipped
```

Reverting the change from `take_along_dim` to `torch.gather` fixes the failure.

**Comment by lezcano** [comment]
It does! Let me put up a forward fix. This means that not all the tests in `torch_np/` are run on CI. @albanD could you give us a hand here? What'd be the best way to add these tests to CI?

**Pull request closed without merging**